src/app/shared/file-editing.ts

```diff
@@ -0,0 +1,92 @@
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+import { Input } from '@angular/core';
+import { Files } from './files';
+import { SourceFile } from './swagger/model/sourceFile';
+
+export class FileEditing extends Files {
+  /**
+   * Toggles edit mode
+   * @return
+   */
+  toggleEdit() {
+    this.editing = !this.editing;
+  }
+
+  /**
+   * Retrieves all dockerfiles from the list of sourcefiles
+   * @param  sourceFiles Array of sourcefiles
+   * @return {Array<SourceFile>}      Array of test parameter files
+   */
+  getDockerFile(sourceFiles: Array<SourceFile>): Array<SourceFile> {
+    return sourceFiles.filter(
+      sourcefile => sourcefile.type === 'DOCKERFILE');
+  }
+
+  /**
+   * Retrieves all descriptor files from the list of sourcefiles
+   * @param  sourceFiles Array of sourcefiles
+   * @return  {Array<SourceFile>}     Array of descriptor files
+   */
+  getDescriptorFiles(sourceFiles: Array<SourceFile>): Array<SourceFile> {
+    return sourceFiles.filter(
+      sourcefile => sourcefile.type === 'DOCKSTORE_WDL' || sourcefile.type === 'DOCKSTORE_CWL');
+  }
+
+  /**
+   * Retrieves all test parameter files from the list of sourcefiles
+   * @param  sourceFiles Array of sourcefiles
+   * @return {Array<SourceFile>}      Array of test parameter files
+   */
+  getTestFiles(sourceFiles: Array<SourceFile>): Array<SourceFile> {
+    return sourceFiles.filter(
+      sourcefile => sourcefile.type === 'WDL_TEST_JSON' || sourcefile.type === 'CWL_TEST_JSON');
+  }
+
+  /**
+   * Common code between tools and workflows for saving
+   * @param originalSourceFiles Sourcefiles from previous version (unedited)
+   * @param newSourceFiles Current set of sourcefiles
+   * @return A list of sourcefiles for the version to be added
+   */
+  commonSaveVersion(originalSourceFiles: Array<SourceFile>, newSourceFiles: Array<SourceFile>): Array<SourceFile> {
+    const sourceFilesToDelete = [];
+
+    // Deal with file renames
+    for (const originalSourceFile of originalSourceFiles) {
+      let toDelete = true;
+      for (const newSourceFile of newSourceFiles) {
+        if (newSourceFile.path === originalSourceFile.path) {
+          toDelete = false;
+          break;
+        }
+      }
+
+      if (toDelete) {
+        const sourceFileCopy = originalSourceFile;
+        sourceFileCopy.content = null;
+        sourceFilesToDelete.push(sourceFileCopy);
+      }
+    }
+
+    if (sourceFilesToDelete.length > 0) {
+      newSourceFiles = newSourceFiles.concat(sourceFilesToDelete);
```

Comment:
what happens if the `if` is left out?

Reply:
I think it was some weird issue with concat adding an undefined or a null element to the end.